### PR TITLE
ztimer: fix required_pm_mode initialization

### PR DESCRIPTION
**Pull request by vincent-d**
### Contribution description

This might be flawed, but from my understanding:
-  the `required_pm_mode` should be applied to the 'base' clock (i.e. the periph_timer or periph_rtt clock)
-  'convert' clocks do not have pm requirement

This PR sets `required_pm_mode` on `ZTIMER_USEC_BASE` and `ZTIMER_MSEC_BASE`. Setting this field is also moved before those clocks are used as they can be used in `ztimer_convert_frac_init()`.
I also changed 'convert' clocks init so they set their internal `required_pm_mode` to `ZTIMER_CLOCK_NO_REQUIRED_PM_MODE` instead of nothing (which makes it 0, thus blocking a power mode by error).

As a last minor change, I fixed the initialization of 'adjust' parameter for the `ZTIMER_MSEC` clock as the field was renamed.

### Testing procedure

I tested with our main application and checked that pm was behaving as expected. Before this change, it could never go to sleep.

### CODE REVIEW

**Comment by kaspar030** [comment]
@jue89 I think this change is valid, could you double check?

**Comment by jue89** [comment]
Good catch! I never unblocked PM mode 0 before ... so this bug never showed up on my application.

Code-wise this looks good to me, too.
Unfortunately, I have no Board at hand, currently. I could test tomorrow on a `samr30-xpro`.

**Comment by kfessel** [comment]
#15715 or this one has to be merged first the other has to be modified

my first impression is that #15715 should go first and this should modifie ZTIMER_SEC after rebase

also this PR should investigate if "ZTIMER_CLOCK_NO_REQUIRED_PM_MODE" is a good(safe) default

**Comment by kfessel** [comment]
it is not ztimer(which scale ever) that needs a specific pm but its base timer there for the configuration value should be
(CONFIG_ZTIMER_RTT_REQUIRED_PM_MODE or CONFIG_ZTIMER_RTC_REQUIRED_PM_MODE)

**Comment by kfessel** [comment]
I am also very fine with rebaseing and editing #15715 after this is merged. If this is faster.
The safe default investigation can be part of another PR. (pm is blocked by default therefor the basic default is safe until unblocked).
Maybe the change of the configuration logic even thoug this touches it should be part of an other PR.
I am not sure how to rename the `ZTIMER_USEC_REQUIRED_PM_MODE` maybe it is `ZTIMER_TIM(ER0)_REQUIRED_PM_MODE`

File: sys/ztimer/auto_init.c
Comment (jue89) [inline sys/ztimer/auto_init.c:118]: This section can cause errors during compilation if `CONFIG_ZTIMER_USEC_REQUIRED_PM_MODE` is configured with the wrongly typed value:

```
In file included from /home/jue/Projects/RIOT/sys/ztimer/auto_init.c:48:
/home/jue/Projects/RIOT/sys/ztimer/auto_init.c: In function 'ztimer_init':
/home/jue/Projects/RIOT/sys/ztimer/auto_init.c:115:15: error: format '%i' expects argument of type 'int', but argument 2 has type 'long unsigned int' [-Werror=format=]
  115 |     LOG_DEBUG("ztimer_init(): ZTIMER_USEC setting required_pm_mode to %i\n",
      |               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

This happens, if use the defines made by the cpu vendor.

I'd say the easy solution is to set `ZTIMER_USEC_BASE->required_pm_mode` and then print that value. If I'm familiar enough with GCC and clang this should fix casting issues due to implicit casting.

```suggestion
#  ifdef MODULE_PM_LAYERED
    ZTIMER_USEC_BASE->required_pm_mode = CONFIG_ZTIMER_USEC_REQUIRED_PM_MODE;
    LOG_DEBUG("ztimer_init(): ZTIMER_USEC setting required_pm_mode to "PRIu8"\n",
              ZTIMER_USEC_BASE->required_pm_mode);
#  endif
```

The same goes with the respective MSEC section.

What do you think?
Maybe @kaspar030 has an opinion on this, too?

**Review by jue89 — CHANGES_REQUESTED** [review]
As advertised, I've tested this PR using the `samr30-xpro`. Sorry that this took so long ...

Power-savings are still achieved - so I'm fine with this PR. Just some minor suggestions.

<details>
<summary>Hacked-together test based on the hello-world example</summary>

```diff
diff --git a/examples/hello-world/Makefile b/examples/hello-world/Makefile
index 258d8e9baf..b72257d9ff 100644
--- a/examples/hello-world/Makefile
+++ b/examples/hello-world/Makefile
@@ -12,6 +12,16 @@ RIOTBASE ?= $(CURDIR)/../..
 # development process:
 DEVELHELP ?= 1
 
+USEMODULE += gnrc_netdev_default
+USEMODULE += auto_init_gnrc_netif
+
+USEMODULE += ztimer ztimer_usec ztimer_msec ztimer_periph_rtt ztimer_xtimer_compat
+USEMODULE += pm_layered
+
+CFLAGS += '-DCONFIG_ZTIMER_MSEC_REQUIRED_PM_MODE=0'
+CFLAGS += '-DCONFIG_ZTIMER_USEC_REQUIRED_PM_MODE=1'
+CFLAGS += '-DPM_BLOCKER_INITIAL=0x00000000'
+
 # Change this to 0 show compiler invocation lines by default:
 QUIET ?= 1
 
diff --git a/examples/hello-world/main.c b/examples/hello-world/main.c
index f51bf8c0a0..fd915fb6b5 100644
--- a/examples/hello-world/main.c
+++ b/examples/hello-world/main.c
@@ -21,6 +21,14 @@
 
 #include <stdio.h>
 
+#include "net/netif.h"
+#include "ztimer.h"
+#include "timex.h"
+
+static void callback (void * arg) {
+       puts((char*) arg);
+}
+
 int main(void)
 {
     puts("Hello World!");
@@ -28,5 +36,20 @@ int main(void)
     printf("You are running RIOT on a(n) %s board.\n", RIOT_BOARD);
     printf("This board features a(n) %s MCU.\n", RIOT_MCU);
 
+    /* The SAM-R30 board boots with its netif enabled and draws several mA of current. */
+    /* Send it to sleep mode ... */
+    netif_t *iface = netif_iter(NULL);
+    netopt_state_t state = NETOPT_STATE_SLEEP;
+    netif_set_opt(iface, NETOPT_STATE, 0, &state, sizeof(netopt_state_t));
+
+    /* 1. Run a callback after 3s */
+    static ztimer_t cb_timer = {.callback = callback, .arg = "Hello World"};
+    ztimer_set(ZTIMER_USEC, &cb_timer, 3 * US_PER_SEC);
+
+    /* 2. Sleep the current thread for 60s */
+    ztimer_sleep(ZTIMER_MSEC, 60 * MS_PER_SEC);
+
+    while (true) {}
+
     return 0;
 }
```

</details>

File: sys/ztimer/auto_init.c
Comment (maribu) [reply sys/ztimer/auto_init.c:118]: If you apply this, please use the past tense in the debug message. (setting -> set)

**Comment by kfessel** [comment]
@vincent-d: would you please check if current #15715 also fixes this.
I moved the PM_MODE configuration to the ZTIMER hardware  bases

**Comment by fjmolinas** [comment]
ping @vincent-d :)

**Comment by kfessel** [comment]
@vincent-d:  #16172 should have solved fix this.

**Comment by vincent-d** [comment]
@kfessel sorry for the long delay. Indeed, it should fix the initialization order issue. But if I'm not mistaken, the convert clock pm mode is still not initialized, though.

**Comment by kfessel** [comment]
@vincent-d: I think we don't need to init the convert_clocks (highlevel like *SEC) pm_mode_block if we init the base clocks (rtt timer or rtc) pm_mode_block

**Comment by maribu** [comment]
> I also changed 'convert' clocks init so they set their internal `required_pm_mode` to `ZTIMER_CLOCK_NO_REQUIRED_PM_MODE` instead of nothing (which makes it 0, thus blocking a power mode by error).

This makes a lot of sense to me. I didn't check to confirm that this does actually cause issues. But IMO purely virtual clocks should use `ZTIMER_CLOCK_NO_REQUIRED_PM_MODE` just because this is the most sensible value there - so even if there is currently no code relying on their value being sensible.

**Comment by maribu** [comment]
@vincent-d: Care to rebase and update?

**Comment by kfessel** [comment]
```
ztimer/convert.c
ztimer/convert_frac.c
ztimer/convert_muldiv64.c
ztimer/convert_shift.c
```
need their init fuctions 
```
super.super.block_pm_mode = ZTIMER_CLOCK_NO_REQUIRED_PM_MODE,
```

it just worked cause i never activated pm_mode 0

**Comment by fjmolinas** [comment]
Any updates here?

**Comment by fjmolinas** [comment]
> ```
> ztimer/convert.c
> ztimer/convert_frac.c
> ztimer/convert_muldiv64.c
> ztimer/convert_shift.c
> ```
> 
> need their init fuctions
> 
> ```
> super.super.block_pm_mode = ZTIMER_CLOCK_NO_REQUIRED_PM_MODE,
> ```
> 
> it just worked cause i never activated pm_mode 0

@kfessel can you provide a patch for what you identified is missing here?

**Comment by kfessel** [comment]
see #16573

**Comment by fjmolinas** [comment]
> see #16573

This PR is now addressed with #16573 being in.

**Pull request closed without merging**
